universe/Condition.cpp

```diff
@@ -9955,6 +9955,203 @@ unsigned int Not::GetCheckSum() const {
     TraceLogger() << "GetCheckSum(Not): retval: " << retval;
     return retval;
 }
+///////////////////////////////////////////////////////////
+// TopmostMatches
+///////////////////////////////////////////////////////////
+TopmostMatches::TopmostMatches(std::vector<std::unique_ptr<ConditionBase>>&& operands) :
+    ConditionBase(),
+    m_operands(std::move(operands))
+{}
+
+bool TopmostMatches::operator==(const ConditionBase& rhs) const {
+    if (this == &rhs)
+        return true;
+    if (typeid(*this) != typeid(rhs))
+        return false;
+
+    const TopmostMatches& rhs_ = static_cast<const TopmostMatches&>(rhs);
+
+    if (m_operands.size() != rhs_.m_operands.size())
+        return false;
+    for (unsigned int i = 0; i < m_operands.size(); ++i) {
+        CHECK_COND_VREF_MEMBER(m_operands.at(i))
+    }
+
+    return true;
+}
+
+void TopmostMatches::Eval(const ScriptingContext& parent_context, ObjectSet& matches,
+               ObjectSet& non_matches, SearchDomain search_domain/* = NON_MATCHES*/) const
+{
+    std::shared_ptr<const UniverseObject> no_object;
+    ScriptingContext local_context(parent_context, no_object);
+
+    if (m_operands.empty()) {
+        ErrorLogger() << "TopmostMatches::Eval given no operands!";
+        return;
+    }
+    for (auto& operand : m_operands) {
+        if (!operand) {
+            ErrorLogger() << "TopmostMatches::Eval given null operand!";
+            return;
+        }
+    }
+
+    // probably i need to figure out which condition matches first
+    // by searching both domains
+    // ... or actually i can propagate the search domain and look for the first match, probably i can
+    // Expressions like  'NOT TopmostMatches [ A B C ]' matches all candidates which do not match the topmost matching condition
+    // which means the same as TopmostMatches [ (Not A) (Not B) (Not C) ]
+    // TODO use logic like in base eval for both domains
+    if (search_domain == NON_MATCHES) {
+        // check all operand conditions on all objects in the non matches set
+        // stop if the operand condition has at least one non match
```

Comment:
Similarly here, I think that if you have `Or [Ship TopmostMatches [Ship Planet]]` and you apply it to an initial input set non_matches containing `{ship 1, planet 2}`, it should match just `{ship 1}`...?

Reply:
Yes, that need fixing

Reply:
fixed